contracts/Controller.sol

```diff
@@ -1,12 +1,297 @@
+    //Call Otoken(args.asset).burn(msg.sender, args.amount)
+    function _burnOtoken (Actions.BurnArgs args) internal {
+
+    }
+
+    //High Level: Otoken holders can withdraw cash value with their oToken
+    function _exercise (Actions.ExerciseArgs args) internal {
+        //Ensure the option has expired
+        //Check priceIsFinalized()
+        //Check oToken.isLocked is false
+        //Amount: 1000000000000000000 // 1 * 10 ^ 18 (no need to scale)
+        //Call MarginCalculator.getExpiredCashValue(oToken, denominated = USDC)
+        //Example cashValue for 1 oToken (option unit): 10.95 USDC, the amount here would be
+        //10 950000 // 10.95 * 10^6. Already convert to USDC decimals
+        //Calculate payoutPer10e18Otoken = cashValue * amount
+        //Payout amount:
+        //High level: 1 oToen * 10.95 usd = 10.95 USDC = 10950000 USDC base unit.
+        //payoutInUSDC =
+        //amount * payoutPer10e18Otoken
+        //(10 * 10 ^18).mul (10.95 * 10^6) .div (10^18)
+        //Call oToken.burn(amount)
+        //MarginPool.transferToUser(payoutInUSDC)
+    }
+
+    //High Level: Clean up a vault and get back collateral after expiry.
+    function _settleVault (Actions.SettleVaultArg args) internal {
+        //Ensure the short oToken has expired
+        //Check priceIsFinalized()
+        //Check oToken.isLocked is false
+        //amountToTakeout = MarginCalculator.getMarginRequirement(vault, collateral)
+        //Burn long oTokens
+        //Call vault.clearVault()
+        //MarginPool.transferToUser(collateral, payout)
+    }
+
+
+    //High Level: call arbitrary smart contract
+    function _call (Actions.CallArgs args) internal {
+        //Check whitelistModule.isWhitelistCallDestination(args.address)
+        //Call args.address with args.data
+    }
+
+    /**
+    * @dev High Level: Checks if the transaction sender is the operator of the owner’s vault
+    * @param _owner The owner of the relevant vault.
+    * @param _sender DO WE NEED THIS?
+    * @return A boolean which is true if and only if the sender is the operator of the owner's vault.
+    */
+    function isOperator (address _owner, address _sender) public view returns (type bool) {
+        //TODO: do we need the second arg, shouldn't it be msg.sender?!
+        //    Return operators[owner][sender]
+        //    setProductUnderlyingPrice (address _otoken, uint256 _roundsBack, ) external
+        //    Get oToken.strike, oToken.underlying
+    }
 
-}
+    /**
+    * @dev the oracle sets the price of the underlying of an otoken.
+    //TODO: why are we setting price of individual otokens, shouldn't we set for an entire batch at once?
+    */
+    function setProductUnderlyingPrice (address _otoken, uint256 _roundsBack, ) external {
+    //TODO: IS THERE SUPPOSED TO BE ANOTHER param?
+        //TODO: why are we setting price of individual otokens, shouldn't we set for an entire batch at once?
```

Comment:
We are setting price per batch here, we just pass Otoken as argument, to hash and get the batch and use that.